contracts/igp/overhead-igp/tests/harness.rs

```diff
@@ -0,0 +1,280 @@
+            "../interchain-gas-paymaster-test/out/debug/interchain-gas-paymaster-test-storage_slots.json".to_string(),
+            vec![],
+        )),
+    )
+    .await
+    .unwrap();
+
+    let test_igp = TestInterchainGasPaymaster::new(test_igp_id.clone(), wallet.clone());
+
+    let overhead_igp_configurables =
+        OverheadIgpConfigurables::default().set_INNER_IGP_ID(Bits256(test_igp_id.hash().into()));
+    let overhead_igp_id = Contract::deploy(
+        "./out/debug/overhead-igp.bin",
+        &wallet,
+        DeployConfiguration::default()
+            .set_storage_configuration(StorageConfiguration::new(
+                "./out/debug/overhead-igp-storage_slots.json".to_string(),
+                vec![],
+            ))
+            .set_configurables(overhead_igp_configurables),
+    )
+    .await
+    .unwrap();
+
+    let overhead_igp = OverheadIgp::new(overhead_igp_id.clone(), wallet.clone());
+
+    // Set the destination gas overhead for the test destination domain
+    let owner_wallet = initial_owner_wallet(&wallet).await.unwrap();
+    overhead_igp
+        .with_wallet(owner_wallet)
+        .unwrap()
+        .methods()
+        .set_destination_gas_overheads(vec![GasOverheadConfig {
+            domain: TEST_DESTINATION_DOMAIN,
+            gas_overhead: TEST_GAS_OVERHEAD_AMOUNT,
+        }])
+        .call()
+        .await
+        .unwrap();
+
+    (overhead_igp, test_igp)
+}
+
+async fn initial_owner_wallet(funder: &WalletUnlocked) -> Result<WalletUnlocked> {
+    funded_wallet_with_private_key(funder, INTIAL_OWNER_PRIVATE_KEY).await
+}
+
+#[tokio::test]
+async fn test_inner_igp_set() {
+    let (overhead_igp, test_igp) = get_contract_instances().await;
+
+    // Sanity check that the inner IGP is set
+    let inner_igp_id = overhead_igp
+        .methods()
+        .inner_igp()
+        .simulate()
+        .await
+        .unwrap()
+        .value;
+    assert_eq!(inner_igp_id, Bits256(test_igp.contract_id().hash().into()));
+}
+
+// ============ pay_for_gas ============
+
+#[tokio::test]
+async fn test_pay_for_gas() {
+    let (overhead_igp, test_igp) = get_contract_instances().await;
+
+    let message_id = Bits256::from_hex_str(TEST_MESSAGE_ID).unwrap();
+    let refund_address = Identity::Address(Address::from_str(TEST_REFUND_ADDRESS).unwrap());
+
+    let payment = 69;
+
+    let call = overhead_igp
+        .methods()
+        .pay_for_gas(
+            message_id,
+            TEST_DESTINATION_DOMAIN,
+            TEST_GAS_AMOUNT,
+            refund_address.clone(),
+        )
+        // Pay 1 base asset
+        .call_params(
+            CallParameters::default()
+                .set_asset_id(BASE_ASSET_ID)
+                .set_amount(payment),
+        )
+        .unwrap()
+        .estimate_tx_dependencies(Some(5))
+        .await
+        .unwrap()
+        .call()
+        .await
+        .unwrap();
+
+    // Check that the inner IGP was called with the correct parameters
```

Comment:
clever, wonder if generalizable to something like `vm.expectCall`

Reply:
Yeah I did a poke around to try to find a way to get something like this by using the receipts from a transaction, but because they just pass around pointers unfortunately a more bespoke setup like this was needed
Hopefully they add some tooling for this at some point